### DISCUSSION TITLE
Symmetry and origin help.

### DISCUSSION THREAD

**Discussion by ctchavez6** [PyAbel/PyAbel#362] [General]
I am trying to use the PyAble transform on a Gaussian profile that I make in numpy. I run the forward transform with Hansen-Law. Then I do the inverse basex transform on the output from the forward transform and should get the original Gaussian profile out, but the result is non-zero and has structure.

I have attached my code which plots the original profile, the output of the forward transform, the output of the inverse transform, and the subtraction of the original and the inverse. 

I notice an issue in the plane x = 2.5, which is half way through my linspace.

I am not sure how to verify that I am setting the origin or the axis of symmetry correctly.

The concept is that I want to simulate an image of a gas jet being shot parallel to the plane of the camera and use PyAble to reconstruct the profile of the gas jet. When  I plot the forward transform it seems like half of the image is lost/changed. 
```python
import plotly.graph_objects as go
import numpy as np
import matplotlib.pyplot as plt
from matplotlib import cm
from matplotlib.ticker import LinearLocator
from scipy.stats import norm
import abel

# Mean = w, SD = x.
# r = 100
s = 0.01 # grid resolution, bigger is faster
W = 0
coll = 0.2 # jet collimation, 0.1-5, smaller = collimated
x = np.arange(0.01, 5, s)
y = np.arange(-10, 10, s)
X, Y = np.meshgrid(x, y)
Z = norm.pdf(Y, W, coll*X)  # Normal Probability density function at x of the given RV.

original = Z
forward_abel = abel.Transform(original, direction='forward', method='hansenlaw', symmetry_axis=1,origin = 'slice').transform
inverse_abel = abel.Transform(forward_abel, direction='inverse', method='basex',symmetry_axis=1).transform


fig = plt.figure( figsize=(20, 8))

# First Plot
ax = fig.add_subplot(1 ,4 ,1, projection='3d')
surf = ax.plot_surface(X, Y, Z, cmap='gnuplot',
                       linewidth=1, antialiased=False)
ax.set_title('original (x,y)')
ax.set_xlabel('X')
ax.set_ylabel('Y')
ax.set_zlabel('Z')
ax.set_ylim(-10, 10)
ax.set_zlim(0, 5)
ax.yaxis.set_major_locator(LinearLocator(5))
ax.yaxis.set_major_formatter('{x:.0f}')
ax.zaxis.set_major_locator(LinearLocator(6))
ax.zaxis.set_major_formatter('{x:.0f}')
ax.w_xaxis.set_pane_color((0.5, 0.5, 0.5, 0.25))
ax.w_yaxis.set_pane_color((0.5, 0.5, 0.5, 0.25))
ax.w_zaxis.set_pane_color((0.5, 0.5, 0.5, 0.25))
fig.colorbar(surf, shrink=0.4, aspect=20) # colorbar

# second plot
ax = fig.add_subplot(1 ,4 ,2)
ax.imshow(forward_abel, clim=(0, np.max(forward_abel) * 0.6), origin='lower')
ax.set_title('Forward Abel')

# third plot
ax = fig.add_subplot(1 ,4 ,3, projection ='3d')
ax.set_title('Inverse Abel')
surf2 = ax.plot_surface(X, Y, inverse_abel, cmap='gnuplot', linewidth=1, antialiased=False)
ax.set_xlabel('X')
ax.set_ylabel('Y')
ax.set_zlabel('Density')
ax.set_ylim(-10, 10)
ax.set_zlim(0, 5)
ax.w_xaxis.set_pane_color((0.5, 0.5, 0.5, 0.25))
ax.w_yaxis.set_pane_color((0.5, 0.5, 0.5, 0.25))
ax.w_zaxis.set_pane_color((0.5, 0.5, 0.5, 0.25))
fig.colorbar(surf2, shrink=0.4, aspect=20) # colorbar


# fourth plot
ax = fig.add_subplot(1 ,4 ,4, projection ='3d')
ax.set_title('Inverse Abel')
surf3 = ax.plot_surface(X, Y, np.subtract(inverse_abel,original), cmap='gnuplot', linewidth=1, antialiased=False)
ax.set_xlabel('X')
ax.set_ylabel('Y')
ax.set_zlabel('Z')
ax.set_ylim(-10, 10)
ax.set_zlim(0, 5)
ax.w_xaxis.set_pane_color((0.5, 0.5, 0.5, 0.25))
ax.w_yaxis.set_pane_color((0.5, 0.5, 0.5, 0.25))
ax.w_zaxis.set_pane_color((0.5, 0.5, 0.5, 0.25))
fig.colorbar(surf3, shrink=0.4, aspect=20) # colorbar


plt.tight_layout()
plt.show()
```

**Comment by stggh** [comment]
@ctchavez6 thanks for using `PyAbel`. 

`PyAbel` expects the image to symmetric about the vertical axis, so your image `Z`:
![gnome-shell-screenshot-d0j3dn](https://user-images.githubusercontent.com/10932229/155425188-03dbef61-024d-47cc-aa76-bdacb09fc4ca.png)
needs to be `Z.T`:
![gnome-shell-screenshot-a0j6ha](https://user-images.githubusercontent.com/10932229/155425358-3c761804-6f25-4e35-8159-33adadcbbbc3.png)

i.e. transpose the image. The `symmetry_axis=1` parameter informs that the top of the image is the same as the bottom. For the transposed image use `symmetry_axis=0` i.e. left size of the image is the same a the right side. This parameter does not inform `PyAbel` of the orientation of the image.

PS: Edited your post (above) to encapsulate python code between code tags

**Comment by ctchavez6** [comment]
@stggh 
Hey and thanks for editing/helping me with my issue. As you can see, I tried using symmetry_axis=1 in both the F and I Able. My understanding of symmetry_axis=1 is that PyAble will transpose the image before and after doing the transform, i.e. should be the same as if I transpose the data myself. i have included the output graph as well. I will try to implement a transposed image as well.

![myplot](https://user-images.githubusercontent.com/86802224/155430729-660d58f5-02a0-4e87-b7ae-8aa461c94cec.png)

**Comment by stggh** [comment]
Here is the a 2D-plot of the intensity difference. Note, here I used `basex` for both directions:
![gnome-shell-screenshot-9xjcej](https://user-images.githubusercontent.com/10932229/155432835-037bdee4-f04e-4ec2-8869-10769b321098.png)

```python
import numpy as np
import matplotlib.pyplot as plt
from scipy.stats import norm
import abel

s = 0.01 # grid resolution, bigger is faster
W = 0
coll = 0.2 # jet collimation, 0.1-5, smaller = collimated
x = np.arange(0.01, 5, s)
y = np.arange(-10, 10, s)
X, Y = np.meshgrid(x, y)
Z = norm.pdf(Y, W, coll*X) # Normal Probability density function at x of the given RV.

Z = Z.T
original=Z

forward_abel = abel.Transform(original, direction='forward',
                              method='basex', symmetry_axis=0).transform
inverse_abel = abel.Transform(forward_abel, direction='inverse',
                              method='basex', symmetry_axis=0).transform

fig, (ax0, ax1, ax2) = plt.subplots(3, 1)
ax0.imshow(Z, vmax=1)
ax0.set_title('Z')
ax1.imshow(inverse_abel, vmax=1)
ax1.set_title('inverse_abel')
diff = Z - inverse_abel
print(f'image difference Z-inverse_abel max={diff.max():.0f} '
      f'min={diff.min():.0f}')
ax2.imshow(diff, vmin=-0.1, vmax=0.1)
ax2.set_title(f'difference min={diff.min():.0f} max={diff.max():.0f}')

plt.tight_layout(h_pad=0.5)
plt.show()
```

**Comment by stggh** [comment]
Actually better without specifying `symmetry_axis=0`:
![gnome-shell-screenshot-bq6umc](https://user-images.githubusercontent.com/10932229/155433217-6ca49074-7c9a-437b-af2b-ee6944de20cf.png)

**Comment by stggh** [comment]
Image needs to be an odd-width and centred.

**Comment by DanHickstein** [comment]
> My understanding of symmetry_axis=1 is that PyAble will transpose the image before and after doing the transform, i.e. should be the same as if I transpose the data myself.

The image always needs to be in the correct orientation. Symmetry_axis controls how if the image is averaged based on up/down or left/right symmetry.
